README.md
<center>Combined Animation Widget</center>

<center>

[![pub package](https://img.shields.io/pub/v/combined_animation.svg)](https://pub.dartlang.org/packages/combined_animation)

</center>

A combined animation widget, contains alignment, opacity, transform, size for enter and leave a widget.

## Features

- ✅ alignment animation.
- ✅ opacity animation.
- ✅ transform animation.
- ✅ size animation.

- ✅ enter animation config.
- ✅ leave animation config.

- ✅ dismiss size animation

## Preview

![preview](preview/preview.gif "preview")

## Getting started

```shell
flutter pub add combined_animation
```

## Usage

```dart
CombinedAnimation(
    state: willRemove?
        ? AnimationType.end
        : AnimationType.start,
    controller: CombinedAnimationController(),
    onEntered: () {
        
    },
    onLeaved: () {
        
    },
    dismissBuilder: (context,size){
        // you can return a size box with size amination to diy dismiss animation
        // or will generate a default one
    }
    onDismiss:(){
        // if you dont remove it
        // you can call controller.enter to show it again
        setState(() {
            doRemove(item);
        });
    },
    config: AnimationConfig.fadeAndZoomIn,
    child: child,
)
```

## Produce

![Produce](preview/produce.png)

CHANGELOG.md

## 0.4.1
* fix

## 0.4.0
* improve config

## 0.3.4
* fix leave

## 0.3.3
* fix a context error

## 0.3.2
* envirment version

## 0.3.1
* reordered callback & state trigger
* fix zero dismiss delay

## 0.3.0
* change leaveBuilder to dismissBuilder
* fix dissmiss to dismiss

## 0.2.0
* Add Controller argument
* Add leaveBuilder to build a dismiss Widget
* Add onInit callback
* change onEntered remove controller argument
* change onLeaved remove size argument

## 0.1.1
* Update example
* Fix static config Type declare

## 0.1.0
* Change: Add controller argument for onEndIn
* Change: arguments name

## 0.0.2

* Fix: ticker.dispose

## 0.0.1

* Init.


test/combined_animation_test.dart
import 'package:flutter/rendering.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:combined_animation/combined_animation.dart';

void main() {
  test('config', () {
    final AnimationConfig config =
        AnimationConfig.fadeIn | AnimationConfig.zoomIn;

    expect(config.startOpacity, 0);
    expect(config.endOpacity, 1);
    expect(config.startTransform?[0], 0.5);
    expect(config.endTransform?[0], 1.0);
  });

  test('snapshot', () {
    final AnimationConfig config =
        AnimationConfig.slideInFrom(const Alignment(0, -3)) |
            AnimationConfig.fadeIn |
            AnimationConfig.zoomIn;

    final align = config.snapshot(0).alignment as Alignment?;
    expect(align?.y, -3);
    final align2 = config.snapshot(0.5).alignment as Alignment?;
    expect(align2?.y, -1.5);
    final align3 = config.snapshot(1).alignment as Alignment?;
    expect(align3?.y, 0);

    expect(config.snapshot(0).opacity, 0);
    expect(config.snapshot(0.5).opacity, 0.5);
    expect(config.snapshot(1).opacity, 1);

    expect(config.snapshot(0).transform?[0], 0);
    expect(config.snapshot(0.5).transform?[0], 0.5);
    expect(config.snapshot(1).transform?[0], 1);
  });
}


example/lib/main.dart
// ignore_for_file: avoid_print

import 'dart:math' as math;

import 'package:combined_animation/combined_animation.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Animation Demo Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final widgets = <Widget>[];
  final configs = <AnimationConfig>[];
  final keys = <ValueKey<int>>[];
  final configList = [
    AnimationConfig.slideInFrom(const Alignment(0, -2)),
    AnimationConfig.slideInFrom(const Alignment(0, 2)),
    AnimationConfig.slideInFrom(const Alignment(-2, 0)),
    AnimationConfig.slideInFrom(const Alignment(2, 0)),
    AnimationConfig.slideAndFadeInFrom(const Alignment(0, -2)),
    AnimationConfig.slideAndFadeInFrom(const Alignment(0, 2)),
    AnimationConfig.slideAndFadeInFrom(const Alignment(-2, 0)),
    AnimationConfig.slideAndFadeInFrom(const Alignment(2, 0)),
    AnimationConfig.zoomIn,
    AnimationConfig.fadeIn,
    AnimationConfig.fadeAndZoomIn,
    AnimationConfig.vFlipIn,
    AnimationConfig.hFlipIn,
    AnimationConfig.zoomIn.copyWith(curve: Curves.bounceOut),
    AnimationConfig.fadeAndZoomIn.copyWith(curve: Curves.bounceOut),
    AnimationConfig.vFlipIn.copyWith(curve: Curves.bounceOut),
    AnimationConfig.hFlipIn.copyWith(curve: Curves.bounceOut),
    AnimationConfig.fadeIn.copyWith(
      startTransform: Matrix4.identity()..rotateZ(math.pi / 2),
      endTransform: Matrix4.identity(),
      curve: Curves.bounceOut,
    )
  ];

  final controller = ScrollController();

  /// generate a new animation item
  void _incrementCounter() {
    setState(() {
      final index = widgets.length;
      widgets.add(Container(
        width: 100.0,
        height: 40.0,
        decoration: BoxDecoration(
          gradient: LinearGradient(
            begin: Alignment.topLeft,
            end: Alignment.bottomRight,
            colors: [
              Colors.primaries[index % 18],
              Colors.primaries[(index + 1) % 18],
              Colors.primaries[(index + 2) % 18],
            ],
          ),
          borderRadius: const BorderRadius.all(Radius.circular(10)),
        ),
      ));
      configs.add(configList[index % configList.length]);
      keys.add(ValueKey(index == 0 ? 0 : keys.last.value + 1));
    });
    WidgetsBinding.instance.scheduleFrameCallback((timeStamp) {
      controller.animateTo(
        controller.position.maxScrollExtent,
        duration: const Duration(milliseconds: 200),
        curve: Curves.easeIn,
      );
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SafeArea(
        child: Column(
          children: [
            const StateDemoWidget(),
            const ControlDemoWidget(),
            Expanded(
              child: ListView.builder(
                controller: controller,
                itemCount: widgets.length,
                padding: const EdgeInsets.only(bottom: 80),
                findChildIndexCallback: (key) {
                  final index = keys.indexOf(key as ValueKey<int>);
                  return index > -1 ? index : null;
                },
                itemBuilder: (context, index) {
                  return AnimateItem(
                    key: keys[index],
                    animate: configs[index],
                    onDismiss: () {
                      setState(() {
                        widgets.removeAt(index);
                        configs.removeAt(index);
                        keys.removeAt(index);
                      });
                    },
                    child: widgets[index],
                  );
                },
              ),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'AddItem',
        child: const Icon(Icons.add),
      ),
    );
  }
}

/// Generate an Anitate Widget control by button
class AnimateItem extends StatefulWidget {
  const AnimateItem({
    Key? key,
    required this.child,
    required this.animate,
    this.onDismiss,
    this.height = 60.0,
  }) : super(key: key);

  final Widget child;

  final VoidCallback? onDismiss;

  final double height;

  final AnimationConfig animate;

  @override
  State<AnimateItem> createState() => _AnimateItemState();
}

class _AnimateItemState extends State<AnimateItem> {
  bool isDissmissing = false;
  bool isLeave = false;
  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        ConstrainedBox(
          constraints: isDissmissing
              ? BoxConstraints.loose(Size.infinite)
              : BoxConstraints.tight(Size.fromHeight(widget.height)),
          child: Center(
            child: CombinedAnimation(
              state: isLeave ? AnimationType.end : AnimationType.start,
              onLeaved: () {
                setState(() {
                  isDissmissing = true;
                });
              },
              onDismiss: widget.onDismiss,
              config: widget.animate,
              child: widget.child,
            ),
          ),
        ),
        if (!isLeave)
          Positioned(
            right: 8,
            top: 0,
            bottom: 0,
            child: Center(
              child: OutlinedButton.icon(
                style: OutlinedButton.styleFrom(
                  shape: const StadiumBorder(),
                ),
                label: const Text('delete'),
                onPressed: () {
                  if (isLeave) return;
                  setState(() {
                    isLeave = true;
                  });
                },
                icon: const Icon(Icons.remove),
              ),
            ),
          ),
      ],
    );
  }
}

class StateDemoWidget extends StatefulWidget {
  const StateDemoWidget({Key? key}) : super(key: key);

  @override
  State<StateDemoWidget> createState() => _StateDemoWidgetState();
}

class _StateDemoWidgetState extends State<StateDemoWidget> {
  AnimationType? state = AnimationType.start;
  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        Container(
          alignment: Alignment.center,
          height: 60,
          child: CombinedAnimation(
            state: state,
            config: AnimationConfig.vFlipIn,
            child: Container(
              width: 100.0,
              height: 40.0,
              decoration: BoxDecoration(
                gradient: LinearGradient(
                  begin: Alignment.topLeft,
                  end: Alignment.bottomRight,
                  colors: [
                    Colors.primaries[0][200]!,
                    Colors.primaries[1][200]!,
                    Colors.primaries[2][200]!,
                  ],
                ),
                borderRadius: const BorderRadius.all(Radius.circular(10)),
              ),
            ),
          ),
        ),
        const Positioned(
          left: 8,
          top: 0,
          bottom: 0,
          child: Align(
            alignment: Alignment.centerLeft,
            child: Text(
              'Control by state',
            ),
          ),
        ),
        Positioned(
          right: 8,
          top: 0,
          bottom: 0,
          child: Center(
            child: OutlinedButton.icon(
              style: OutlinedButton.styleFrom(
                shape: const StadiumBorder(),
              ),
              label: Text(
                state == null
                    ? 'Show'
                    : state == AnimationType.start
                        ? 'Hide'
                        : 'Reset',
              ),
              onPressed: () {
                final cState = (state?.index ?? -1) + 1;

                setState(() {
                  state = cState > 1 ? null : AnimationType.values[cState];
                });
              },
              icon: const Icon(Icons.animation),
            ),
          ),
        ),
      ],
    );
  }
}

/// control animation by controller
class ControlDemoWidget extends StatefulWidget {
  const ControlDemoWidget({Key? key}) : super(key: key);

  @override
  State<ControlDemoWidget> createState() => _ControlDemoWidgetState();
}

class _ControlDemoWidgetState extends State<ControlDemoWidget> {
  /// control by controller
  late CombinedAnimationController caController = CombinedAnimationController()
    ..addListener(() {
      print(
          '${caController.state} ${caController.isEntered} ${caController.isLeaved}');
      setState(() {});
    });

  @override
  void dispose() {
    caController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        Container(
          alignment: Alignment.center,
          height: 60,
          child: CombinedAnimation(
            config: AnimationConfig.vFlipIn,
            controller: caController,
            isControlled: true,
            child: Container(
              width: 100.0,
              height: 40.0,
              decoration: BoxDecoration(
                gradient: LinearGradient(
                  begin: Alignment.topLeft,
                  end: Alignment.bottomRight,
                  colors: [
                    Colors.primaries[9][200]!,
                    Colors.primaries[10][200]!,
                    Colors.primaries[11][200]!,
                  ],
                ),
                borderRadius: const BorderRadius.all(Radius.circular(10)),
              ),
            ),
          ),
        ),
        const Positioned(
          left: 8,
          top: 0,
          bottom: 0,
          child: Align(
            alignment: Alignment.centerLeft,
            child: Text(
              'Control by controller',
            ),
          ),
        ),
        Positioned(
          right: 8,
          top: 0,
          bottom: 0,
          child: Center(
            child: OutlinedButton.icon(
              style: OutlinedButton.styleFrom(
                shape: const StadiumBorder(),
              ),
              label: Text((caController.isLeaved)
                  ? 'Reset'
                  : (caController.isEntered)
                      ? 'Hide'
                      : 'Show'),
              onPressed: () {
                print(caController.isEntered);
                if (caController.isEntered) {
                  caController.leave();
                } else if (caController.isLeaved) {
                  caController.init();
                  setState(() {});
                } else {
                  caController.enter();
                }
              },
              icon: const Icon(Icons.animation),
            ),
          ),
        ),
      ],
    );
  }
}


lib/src/combined.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

import 'config.dart';

const _minFrameDelay = 16;

/// Only allow user to start or stop
enum AnimationType {
  start,
  end,
}

/// Combine state of [CombinedAnimation]
enum AnimationState {
  init,
  beginEnter,
  endEnter,
  beginLeave,
  endLeave,
  dismiss,
}

const _defaultDuration = Duration(milliseconds: 300);

/// A combined animation to show in or hide out a widget
class CombinedAnimation extends StatefulWidget {
  const CombinedAnimation({
    Key? key,
    required this.config,
    AnimationConfig? leaveConfig,
    required this.child,
    AnimationType? state = AnimationType.start,
    this.dismissDuration = _defaultDuration,
    this.dismissCurve = Curves.easeOut,
    this.controller,
    this.onInit,
    this.onEntered,
    this.onLeaved,
    this.onDismiss,
    this.dismissBuilder,
    this.autoSlide = true,
    this.isControlled = false,
  })  : leaveConfig = leaveConfig ?? ~config,
        state = state == AnimationType.start
            ? AnimationState.beginEnter
            : state == AnimationType.end
                ? AnimationState.beginLeave
                : AnimationState.init,
        super(key: key);

  /// Enter animation config
  final AnimationConfig config;

  /// Leave animation config. defaults to flip [config]
  final AnimationConfig leaveConfig;

  /// State to pass in
  final AnimationState state;

  /// pass a controller to attach state
  final CombinedAnimationController? controller;

  /// Callback to build a size transation to dismiss
  final Widget? Function(Size?)? dismissBuilder;

  /// Callback when state inited
  final VoidCallback? onInit;

  /// Callback when show in animation is complete
  final VoidCallback? onEntered;

  /// Callback when leave animation is complete
  final VoidCallback? onLeaved;

  /// Callback when size is Zero
  final VoidCallback? onDismiss;

  /// The child will be animate
  final Widget child;

  /// Whether should auto slide up
  final bool autoSlide;

  /// If is controlled, State will not update state from widget
  final bool isControlled;

  /// animation duration to dismiss
  final Duration dismissDuration;

  /// animation curve to dismiss
  final Curve dismissCurve;

  @override
  State<CombinedAnimation> createState() => _CombinedAnimationState();
}

/// AnimationController to control Animation State
class CombinedAnimationController extends ChangeNotifier {
  CombinedAnimationController();

  _CombinedAnimationState? _state;
  bool _leaved = false;

  AnimationState get state => _state?.state ?? AnimationState.init;

  bool get isEntered => _state?.state == AnimationState.endEnter;

  bool get isLeaved =>
      (_state?.state.index ?? 0) >= AnimationState.endLeave.index;

  void _attach(_CombinedAnimationState state) {
    _state = state;
  }

  void _stateChanged() {
    if (_state != null) notifyListeners();
  }

  /// reset
  void init() {
    _state?.init();
  }

  /// stop animation
  void stop() {
    _state?.stop();
  }

  /// start enter animation
  void enter({Duration? duration}) {
    if (_leaved) {
      if (_state != null) {
        _leaved = false;
      }
      return;
    }
    _state?.enter(duration: duration);
  }

  /// start leave animation
  void leave({Duration? duration}) {
    if (_state == null) _leaved = true;
    _state?.leave(duration: duration);
  }

  @override
  void dispose() {
    super.dispose();
    _state = null;
  }
}

class _CombinedAnimationState extends State<CombinedAnimation>
    with SingleTickerProviderStateMixin<CombinedAnimation> {
  late AnimationSnapshot snapshot;

  AnimationState state = AnimationState.init;

  late final animation = AnimationController(vsync: this);

  Size? size;

  bool get isEnter =>
      state.index >= AnimationState.beginEnter.index &&
      state.index <= AnimationState.endEnter.index;

  bool get isLeave =>
      state.index >= AnimationState.beginLeave.index &&
      state.index <= AnimationState.endLeave.index;

  @override
  void initState() {
    super.initState();
    snapshot = widget.config.snapshot(0);
    state = widget.state;

    animation.addListener(_onAnimation);

    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      if (state == AnimationState.beginEnter) {
        enter();
      }
      widget.onInit?.call();
    });

    widget.controller?._attach(this);
  }

  void init() {
    stop();
    state = AnimationState.init;
    setState(() {
      snapshot = widget.config.snapshot(0);
    });
    widget.controller?._stateChanged();
  }

  void stop({bool canceled = true}) {
    if (animation.isAnimating) {
      animation.stop(canceled: canceled);
    }
  }

  void enter({Duration? duration}) {
    if (!mounted) return;
    if (state.index < AnimationState.endLeave.index &&
        state.index > AnimationState.beginEnter.index) {
      return;
    }
    state = AnimationState.beginEnter;
    widget.controller?._stateChanged();
    animation
        .animateTo(
      1,
      duration: duration ?? widget.config.duration ?? _defaultDuration,
      curve: widget.config.curve ?? Curves.easeIn,
    )
        .whenComplete(() {
      if (mounted) {
        size = context.size;
      }
      state = AnimationState.endEnter;
      widget.controller?._stateChanged();
      widget.onEntered?.call();
    });
  }

  void leave({Duration? duration}) {
    if (!mounted) return;
    if (state.index >= AnimationState.beginLeave.index) return;
    state = AnimationState.beginLeave;
    size = context.size;
    widget.controller?._stateChanged();
    animation
        .animateTo(
      0,
      duration: duration ?? widget.leaveConfig.duration ?? _defaultDuration,
      curve: widget.leaveConfig.curve ?? Curves.easeOut,
    )
        .whenComplete(() {
      state = AnimationState.endLeave;
      widget.controller?._stateChanged();
      widget.onLeaved?.call();

      WidgetsBinding.instance.scheduleFrameCallback((timeStamp) {
        if (size == null ||
            widget.dismissDuration.inMilliseconds < _minFrameDelay) {
          state = AnimationState.dismiss;
          widget.controller?._stateChanged();
          widget.onDismiss?.call();
        } else {
          size = null;
          if (mounted) {
            setState(() {});
          }

          Future.delayed(
            Duration(
              milliseconds:
                  widget.dismissDuration.inMilliseconds + _minFrameDelay,
            ),
          ).then((value) {
            state = AnimationState.dismiss;
            widget.controller?._stateChanged();
            widget.onDismiss?.call();
          });
        }
      });
    });
  }

  @override
  void dispose() {
    animation.removeListener(_onAnimation);
    animation.dispose();
    super.dispose();
  }

  /// Update state. maybe animate to hide out or reset or restart
  @override
  void didUpdateWidget(covariant CombinedAnimation oldWidget) {
    super.didUpdateWidget(oldWidget);
    widget.controller?._attach(this);
    if (widget.isControlled) return;
    if (widget.state == AnimationState.beginLeave) {
      if (state == AnimationState.endEnter) {
        WidgetsBinding.instance.addPostFrameCallback((_) {
          leave();
        });
      }
    } else if (widget.state == AnimationState.beginEnter) {
      enter();
    } else if (widget.state == AnimationState.init &&
        !animation.isAnimating &&
        state == AnimationState.dismiss) {
      init();
    }
  }

  void _onAnimation() {
    setState(() {
      snapshot = state.index > AnimationState.endEnter.index
          ? widget.leaveConfig.snapshot(1 - animation.value)
          : widget.config.snapshot(animation.value);
    });
  }

  @override
  Widget build(BuildContext context) {
    if (state.index >= AnimationState.endLeave.index &&
        widget.autoSlide &&
        (widget.dismissBuilder != null ||
            !(widget.config.hasSize || widget.leaveConfig.hasSize))) {
      return widget.dismissBuilder?.call(size) ??
          AnimatedSize(
            duration: widget.dismissDuration,
            curve: widget.dismissCurve,
            alignment: Alignment.topCenter,
            child: SizedBox(
              width: size?.width ?? 0,
              height: size?.height ?? 0,
            ),
          );
    }
    Widget child = widget.child;

    if (widget.config.hasMatrix || widget.leaveConfig.hasMatrix) {
      child = Transform(
        transform: snapshot.transform ?? Matrix4.identity(),
        alignment: Alignment.center,
        child: widget.child,
      );
    }
    if (widget.config.hasSize || widget.leaveConfig.hasSize) {
      child = AnimatedSize(
        duration:
            (isEnter ? widget.config.duration : widget.leaveConfig.duration) ??
                _defaultDuration,
        child: child,
      );
    }
    if (widget.config.hasOpacity || widget.leaveConfig.hasOpacity) {
      child = Opacity(
        opacity: snapshot.opacity ?? 1,
        child: child,
      );
    }

    if (widget.config.hasAlign || widget.leaveConfig.hasAlign) {
      child = Align(
        alignment: snapshot.alignment ?? Alignment.center,
        child: child,
      );
    }

    return child;
  }
}


lib/src/config.dart
import 'dart:math' as math;
import 'dart:ui' show lerpDouble;

import 'package:flutter/widgets.dart';

/// A snapshot of an animation config
class AnimationSnapshot {
  AnimationSnapshot({
    this.transform,
    this.opacity,
    this.alignment,
  });

  final Matrix4? transform;
  final double? opacity;
  final AlignmentGeometry? alignment;

  Map<String, dynamic> toJson() => {
        'transform': transform,
        'opacity': opacity,
        'alignment': alignment,
      };
  @override
  String toString() => toJson().toString();
}

/// Animation config
class AnimationConfig {
  /// Slide in from `start` [Alignment]
  static AnimationConfig slideInFrom(Alignment start) => AnimationConfig(
        startAlign: start,
        endAlign: Alignment.center,
        curve: Curves.easeOutQuad,
      );

  /// Slide out to `end` [Alignment]
  static AnimationConfig slideOutTo(Alignment end) => AnimationConfig(
        startAlign: Alignment.center,
        endAlign: end,
        curve: Curves.easeOutQuad,
      );

  /// From out top to center
  @Deprecated('use expandIn')
  static AnimationConfig get slideIn => slideInFrom(const Alignment(0, -3));

  /// From center to out top
  @Deprecated('use expandIn')
  static AnimationConfig get slideOut => slideOutTo(const Alignment(0, -3));

  @Deprecated('use expandIn')
  static AnimationConfig get slideDown => expandIn;

  /// Expand in
  static const expandIn = AnimationConfig(
    startSize: Size.zero,
    endSize: Size.infinite,
    curve: Curves.easeOutQuad,
  );

  @Deprecated('use shrinkOut')
  static AnimationConfig get slideUp => shrinkOut;

  /// Shrink out
  static const shrinkOut = AnimationConfig(
    startSize: Size.infinite,
    endSize: Size.zero,
    curve: Curves.easeOutQuad,
  );

  /// Slide and fade in from `start` [Alignment]
  static AnimationConfig slideAndFadeInFrom(Alignment start) => AnimationConfig(
        startAlign: start,
        endAlign: Alignment.center,
        startOpacity: 0,
        endOpacity: 1,
        curve: Curves.easeOutQuad,
      );

  /// From out top to center with fade in
  @Deprecated('use slideAndFadeInFrom')
  static AnimationConfig get slideAndFadeIn => slideAndFadeInFrom(
        const Alignment(0, -3),
      );

  /// Slide and fade out from `end` [Alignment]
  static AnimationConfig slideAndFadeOutTo(Alignment end) => AnimationConfig(
        startAlign: Alignment.center,
        endAlign: end,
        startOpacity: 1,
        endOpacity: 0,
        curve: Curves.easeOutQuad,
      );

  /// From center to out top with fade out
  @Deprecated('use slideAndFadeOutTo')
  static AnimationConfig get slideAndFadeOut => slideAndFadeOutTo(
        const Alignment(0, -3),
      );

  /// Fade in
  static const fadeIn = AnimationConfig(
    startOpacity: 0,
    endOpacity: 1,
    curve: Curves.easeOutQuad,
  );

  /// Fade out
  static const fadeOut = AnimationConfig(
    startOpacity: 1,
    endOpacity: 0,
    curve: Curves.easeOutQuad,
  );

  /// Zoom in
  static final zoomIn = AnimationConfig(
    startTransform: Matrix4.identity()..scale(0.0),
    endTransform: Matrix4.identity(),
    curve: Curves.easeOutQuad,
  );

  /// Zoom out
  static final AnimationConfig zoomOut = ~zoomIn;

  /// Zoom in with fade in
  static final fadeAndZoomIn = AnimationConfig(
    startOpacity: 0,
    endOpacity: 1,
    startTransform: Matrix4.identity()..scale(0.5),
    endTransform: Matrix4.identity(),
    curve: Curves.easeOutQuad,
  );

  /// Zoom out with fade out
  static final AnimationConfig fadeAndZoomOut = ~fadeAndZoomIn;

  static final vFlipIn = AnimationConfig(
    startTransform: Matrix4.identity()..rotateX(math.pi / 2),
    endTransform: Matrix4.identity(),
    curve: Curves.easeOutQuad,
  );

  static final AnimationConfig vFlipOut = ~vFlipIn;

  static final hFlipIn = AnimationConfig(
    startTransform: Matrix4.identity()..rotateY(math.pi / 2),
    endTransform: Matrix4.identity(),
    curve: Curves.easeOutQuad,
  );

  static final AnimationConfig hFlipOut = ~hFlipIn;

  /// All paired params must be provided or null.
  const AnimationConfig({
    this.startAlign,
    this.endAlign,
    this.startOpacity,
    this.endOpacity,
    this.startTransform,
    this.endTransform,
    this.startSize,
    this.endSize,
    this.duration,
    this.curve,
  })  : assert(
          (startAlign != null && endAlign != null) ||
              (startAlign == null && endAlign == null),
          'Align animation need start and end non null',
        ),
        assert(
          (startOpacity != null && endOpacity != null) ||
              (startOpacity == null && endOpacity == null),
          'Opacity animation need start and end non null',
        ),
        assert(
          (startTransform != null && endTransform != null) ||
              (startTransform == null && endTransform == null),
          'Transform animation need start and end non null',
        ),
        assert(
          (startSize != null && endSize != null) ||
              (startSize == null && endSize == null),
          'Transform animation need start and end non null',
        );

  /// Quick create an enter config
  AnimationConfig.enter({
    AlignmentGeometry? align,
    double? opacity,
    Matrix4? transform,
    Size? size,
    Duration? duration,
    Curve? curve,
  }) : this(
          startAlign: align,
          endAlign: align == null ? null : Alignment.center,
          startOpacity: opacity,
          endOpacity: opacity == null ? null : 1,
          startTransform: transform,
          endTransform: transform == null ? null : Matrix4.identity(),
          startSize: size,
          endSize: size == null ? null : Size.infinite,
          duration: duration,
          curve: curve,
        );

  /// Quick create a leave config
  AnimationConfig.leave({
    AlignmentGeometry? align,
    double? opacity,
    Matrix4? transform,
    Size? size,
    Duration? duration,
    Curve? curve,
  }) : this(
          startAlign: align == null ? null : Alignment.center,
          endAlign: align,
          startOpacity: opacity == null ? null : 1,
          endOpacity: opacity,
          startTransform: transform == null ? null : Matrix4.identity(),
          endTransform: transform,
          startSize: size == null ? null : Size.infinite,
          endSize: size,
          duration: duration,
          curve: curve,
        );

  /// alignment
  final AlignmentGeometry? startAlign;
  final AlignmentGeometry? endAlign;

  /// opacity
  final double? startOpacity;
  final double? endOpacity;

  /// transform
  final Matrix4? startTransform;
  final Matrix4? endTransform;

  /// size
  final Size? startSize;
  final Size? endSize;

  final Duration? duration;
  final Curve? curve;

  bool get hasAlign => startAlign != null;
  bool get hasOpacity => startOpacity != null;
  bool get hasMatrix => startTransform != null;
  bool get hasSize => startSize != null;

  /// Generate a snapshot of current animation [value]
  AnimationSnapshot snapshot(double value) {
    if (value == 0) {
      return AnimationSnapshot(
        transform: startTransform,
        alignment: startAlign,
        opacity: startOpacity,
      );
    }
    if (value == 1) {
      return AnimationSnapshot(
        transform: endTransform,
        alignment: endAlign,
        opacity: endOpacity,
      );
    }
    Matrix4? matrix4;
    if (hasMatrix) {
      matrix4 = Matrix4.fromList(
        List.generate(
          16,
          (idx) => lerpDouble(startTransform![idx], endTransform![idx], value)!,
        ),
      );
    }
    double? opacity;
    if (hasOpacity) {
      opacity = lerpDouble(startOpacity, endOpacity, value) ?? 1;
    }
    AlignmentGeometry? alignment;
    if (hasAlign) {
      alignment = AlignmentGeometry.lerp(startAlign!, endAlign!, value);
    }

    return AnimationSnapshot(
      transform: matrix4,
      alignment: alignment,
      opacity: opacity,
    );
  }

  /// Creates a copy of this config
  /// but with the given fields replaced with the new values.
  AnimationConfig copyWith({
    AlignmentGeometry? startAlign,
    AlignmentGeometry? endAlign,
    double? startOpacity,
    double? endOpacity,
    Matrix4? startTransform,
    Matrix4? endTransform,
    Size? startSize,
    Size? endSize,
    Duration? duration,
    Curve? curve,
  }) =>
      AnimationConfig(
        startAlign: startAlign ?? this.startAlign,
        endAlign: endAlign ?? this.endAlign,
        startOpacity: startOpacity ?? this.startOpacity,
        endOpacity: endOpacity ?? this.endOpacity,
        startTransform: startTransform ?? this.startTransform,
        endTransform: endTransform ?? this.endTransform,
        startSize: startSize ?? this.startSize,
        endSize: endSize ?? this.endSize,
        duration: duration ?? this.duration,
        curve: curve ?? this.curve,
      );

  /// Combine two config
  AnimationConfig operator |(AnimationConfig other) {
    return AnimationConfig(
      startAlign: startAlign ?? other.startAlign,
      endAlign: endAlign ?? other.endAlign,
      startOpacity: startOpacity ?? other.startOpacity,
      endOpacity: endOpacity ?? other.endOpacity,
      startTransform: startTransform ?? other.startTransform,
      endTransform: endTransform ?? other.endTransform,
      startSize: startSize ?? other.startSize,
      endSize: endSize ?? other.endSize,
      duration: (duration?.compareTo(other.duration ?? Duration.zero) ?? 0) > 0
          ? duration
          : other.duration,
      curve: curve,
    );
  }

  /// Flip config
  AnimationConfig operator ~() {
    return AnimationConfig(
      startAlign: endAlign,
      endAlign: startAlign,
      startOpacity: endOpacity,
      endOpacity: startOpacity,
      startTransform: endTransform,
      endTransform: startTransform,
      startSize: endSize,
      endSize: startSize,
      duration: duration,
      curve: curve?.flipped,
    );
  }

  /// To json
  Map<String, dynamic> toJson() => {
        'startAlign': startAlign,
        'endAlign': endAlign,
        'startOpacity': startOpacity,
        'endOpacity': endOpacity,
        'startTransform': startTransform,
        'endTransform': endTransform,
        'startSize': startSize,
        'endSize': endSize,
        'duration': duration,
        'curve': curve,
      };

  @override
  String toString() => toJson().toString();
}
